internal/stun/server.go

```diff
@@ -0,0 +1,134 @@
+package stun
+
+import (
+	"errors"
+	"github.com/nexodus-io/nexodus/internal/util"
+	"github.com/pion/stun"
+	"go.uber.org/zap"
+	"net"
+	"strconv"
+	"sync"
+)
+
+func ListenAndStart(address string) (*Server, error) {
+	conn, err := net.ListenPacket("udp", address)
+	if err != nil {
+		return nil, err
+	}
+
+	_, port, err := net.SplitHostPort(conn.LocalAddr().String())
+	if err != nil {
+		return nil, err
+	}
+	p, err := strconv.Atoi(port)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &Server{
+		conn: conn,
+		Port: p,
+		server: server{
+			Log: zap.NewNop(),
+		},
+	}
+
+	util.GoWithWaitGroup(&s.wg, func() {
+		err = s.Serve(conn)
+		if err != nil {
+			s.Log.Info("Failed Serve", zap.Error(err))
+		}
+	})
+
+	return s, nil
+}
+
+type Server struct {
+	conn net.PacketConn
+	wg   sync.WaitGroup
+	server
+	Port int
+}
+
+func (s *Server) Close() error {
+	return s.conn.Close()
+}
+func (s *Server) Shutdown() error {
+	err := s.Close()
+	if err != nil {
+		return err
+	}
+	s.wg.Wait()
+	return nil
+}
+
+func ListenAndServe(address string) error {
+	conn, err := net.ListenPacket("udp", address)
+	if err != nil {
+		return err
+	}
+	s := &server{
+		Log: zap.NewNop(),
+	}
+	return s.Serve(conn)
+}
+
+type server struct {
+	Log *zap.Logger
+}
+
+func (s *server) Serve(conn net.PacketConn) error {
+	buf := make([]byte, 1024)
+	response := stun.Message{}
+	request := stun.Message{}
+	fromAddress := stun.XORMappedAddress{}
+	var software = stun.NewSoftware("nexodus")
+	for {
+
+		n, addr, err := conn.ReadFrom(buf)
+		if err != nil {
+			if errors.Is(err, net.ErrClosed) {
+				return nil
+			}
+			return err
+		}
+		requestBytes := buf[:n]
+
+		if !stun.IsMessage(requestBytes) {
+			s.Log.Info("Not a STUN request")
```

Comment:
I'd make this a debug message. Otherwise, it could be a low-key DoS by filling up a device's disk with log messages by sending it a stream of tiny UDP packes